[HLSL][RootSignature] Define and integrate rootsig clang attr and decl

inbelic opened the pull request:
- Defines a new declaration node `HLSLRootSignature` in `DeclNodes.td` that will consist of a `TrailingObjects` of the in-memory construction of the root signature, namely an array of `hlsl::rootsig::RootElement`s

- Defines a new clang attr `RootSignature` which simply holds an identifier to a corresponding root signature declration as above

- Integrate the `HLSLRootSignatureParser` to construct the decl node in `ParseMicrosoftAttributes` and then attach the parsed attr with an identifier to the entry point function declaration.

- Defines the various required declaration methods

- Add testing that the declaration and reference attr are created correctly, and some syntactical error tests.

It was previously proposed that we could have the root elements reference be stored directly as an additional member of the attribute and to not have a seperate root signature decl. In contrast, by defining them seperately as this change proposes, we will allow a unique root signature to have its own declaration in the AST tree. This allows us to only construct a single root signature for all duplicate rootsignature attributes. Having it located directly as a declaration might also prove advantageous when we consider root signature libraries.

Resolves https://github.com/llvm/llvm-project/issues/119011

inbelic committed:
[HLSL][RootSignature] Define and integrate rootsig clang attr and decl

- Defines a new declaration node `HLSLRootSignature` in `DeclNodes.td`
that will hold a reference to an in-memory construction of the root
signature, namely an array of `hlsl::rootsig::RootElement`s

- Defines a new clang attr `RootSignature` which simply holds an
identifier to a corresponding root signature declration as above

It was previously proposed that we could have the root elements
reference be stored directly as an additional member of the attribute
and to not have a seperate root signature decl. In contrast, by defining
them seperately as this change proposes, we will allow a unique root
signature to have its own declaration in the AST tree. This allows us
to only construct a single root signature for all duplicate rootsignature
attributes. Having it located directly as a declaration might also prove
advantageous when we consider root signature libraries.

inbelic committed:
clang format

inbelic commented on clang/lib/AST/TextNodeDumper.cpp:
```diff
@@ -3037,6 +3037,11 @@ void TextNodeDumper::VisitHLSLBufferDecl(const HLSLBufferDecl *D) {
 
+void TextNodeDumper::VisitHLSLRootSignatureDecl(
+    const HLSLRootSignatureDecl *D) {
+  dumpName(D);
```
We could extend this to dump more info about the root elements. For instance, number of elements, or iterate through the types, etc.

llvm-beanz replied:
We should think through how we want to serialize the root signature to text as part of the AST. It might be nice to have something so that we can test/inspect it.

inbelic replied:
Noting:

There exists unit testing to check in-memory representation is constructed correctly
Tests in this pr ensure that the decl and attr are constructed/allocated

So there is technically a small test gap that the constructed root-elements are copied correctly over, which could be demonstrated by adding serialization for testing.

With that said, I think adding serialization to this pr would get out of its scope, so https://github.com/llvm/llvm-project/issues/138025 will track that.

Do we want to block this pr on implementing the serialization of the elements used in the testcase? (I have verified the elements are copied over correctly on a local quick-hack impl of serialization)

inbelic replied:
From design we will block on https://github.com/llvm/llvm-project/issues/138025 such that we can provide sufficient testing.

We will go with an initial approach to just dump all information of the in-memory constructs and we can revisit this to maybe only print explicitly set parameters if this becomes to verbose/hard to debug with.

inbelic commented on clang/test/SemaHLSL/RootSignature-err.hlsl:
```diff
@@ -0,0 +1,11 @@
+
+// expected-error@+2 {{expected ')'}}
+// expected-note@+1 {{to match this '('}}
+[RootSignature("", "")]
```
The location here points to the comma

inbelic commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -949,6 +950,23 @@ void SemaHLSL::emitLogicalOperatorFixIt(Expr *LHS, Expr *RHS,
+    Diag(AL.getLoc(), diag::err_attribute_wrong_number_arguments) << AL << 1;
+    return;
+  }
+
```
Should also validate that we haven't already received a root signature attr

llvm-beanz replied:
We should only allow the attribute to appear more than once if it is the same root signature. Mismatching root signature attributes should be an error (err_disallowed_duplicate_attribute), and duplicate but identical attributes a warning (warn_duplicate_attribute_exact).

bogner commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -5209,6 +5211,92 @@ void Parser::ParseMicrosoftUuidAttributeArgs(ParsedAttributes &Attrs) {
 }
 
+void Parser::ParseMicrosoftRootSignatureAttributeArgs(ParsedAttributes &Attrs) {
+  assert(Tok.is(tok::identifier) && "Not a Microsoft attribute list");
+  IdentifierInfo *RootSignatureIdent = Tok.getIdentifierInfo();
+  assert(RootSignatureIdent->getName() == "RootSignature" &&
+         "Not a Microsoft attribute list");
```
The messages in these asserts aren't relevant to the problem... These should probably say something like "Expected RootSignature identifier"

bogner commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -5209,6 +5211,92 @@ void Parser::ParseMicrosoftUuidAttributeArgs(ParsedAttributes &Attrs) {
+  }
+
+  ArgsVector Args;
+  if (auto Lit = dyn_cast<StringLiteral>(StringResult.get())) {
```
What should happen if this *isn't* a StringLiteral? The code looks like it just silently drops it if that happens currently

inbelic replied:
Updated the process of extracting the `StringLiteral` so that it better maps to the diag reporting it

bogner commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -5209,6 +5211,92 @@ void Parser::ParseMicrosoftUuidAttributeArgs(ParsedAttributes &Attrs) {
+
+      // Create the Root Signature
+      auto *SignatureDecl = HLSLRootSignatureDecl::Create(
+          Actions.getASTContext(), /*FIXME?*/ Actions.CurContext,
```
Leftover FIXME

bogner reviewed (commented):
A couple of minor comments. The direction seems reasonable to me, but it'd be good to get someone more familiar with the Parser and AST to weigh in.

llvm-beanz commented on clang/include/clang/AST/Decl.h:
```diff
@@ -5178,6 +5179,29 @@ class HLSLBufferDecl final : public NamedDecl, public DeclContext {
 };
 
+class HLSLRootSignatureDecl final : public NamedDecl {
+  ArrayRef<llvm::hlsl::rootsig::RootElement> RootElements;
```
This should probably be a `TrailingObjects` instead of an `ArrayRef` so that it is allocated with the Decl (and in adjacent cache-friendly memory).

llvm-beanz reviewed (commented):
A few comments, but I think this is mostly the right direction.

inbelic committed:
review: improve assert messages

inbelic committed:
review: remove FIXME

- this is correct as we do want to add the newly constructed Declaration
into the current decl context, such that we can retrieve it later when
we look it up from the function declaration in the same context

inbelic committed:
review: unify error reporting of having an invalid string literal argument

inbelic committed:
clang-format

inbelic committed:
review: use TrailingObjects to store elements instead of on the context

inbelic committed:
review: update duplicate attrs warning/error

inbelic committed:
self-review: correct retrieved start location

- increment the location by one to better point past the initial "
character of a string literal
- add test that demonstrates an HLSLRootSignature parsing error being
propogated up and it correctly continues parsing after the closed paren
enclosing the string literal

inbelic committed:
self-review: update skip logic to account for context

inbelic committed:
clang-format

inbelic commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -5209,6 +5211,97 @@ void Parser::ParseMicrosoftUuidAttributeArgs(ParsedAttributes &Attrs) {
+      return;
+    }
+
+    // Allocate the root elements onto ASTContext
+    unsigned N = Elements.size();
+    auto RootElements = MutableArrayRef<llvm::hlsl::rootsig::RootElement>(
+        ::new (Actions.getASTContext()) llvm::hlsl::rootsig::RootElement[N], N);
+    for (unsigned I = 0; I < N; ++I)
+      RootElements[I] = Elements[I];
```
Whoops this should be able to get deleted now and we can just pass `Elements` where `RootElements is below

inbelic committed:
rebase: hook-up root elements to invoke there dump

- update ast test to ensure elements are correctly allocated

inbelic committed:
self-review: remove ast allocation

inbelic committed:
clang-format

inbelic commented:
This is rebased onto: https://github.com/llvm/llvm-project/pull/138326 which defines a serialization of the in-memory structures.

As such, the test-case in RootSignatures-AST.hlsl is updated to show that the values are correctly stored in the decl node.

bogner commented on llvm/lib/Frontend/HLSL/HLSLRootSignature.cpp:
```diff
@@ -149,6 +149,29 @@ void DescriptorTableClause::dump(raw_ostream &OS) const {
   OS << ")";
 }
 
+// Helper struct so that we can use the overloaded notation of std::visit
+template <class... Ts> struct OverloadMethods : Ts... {
+  using Ts::operator()...;
+};
+
+template <class... Ts> OverloadMethods(Ts...) -> OverloadMethods<Ts...>;
```
I feel like it'd be simpler to just define a callable with a templated `operator()`:
```c++
namespace {
struct ElementDumper {
  raw_ostream &OS;
  template <typename T> void operator()(const T &Element) const {
    Element.dump(OS);
  }
};
}
```
Then we just need `std::visit(ElementDumper{OS}, Element)` later.

inbelic committed:
review: remove forced copy

inbelic committed:
review: simplify overloaded notation

inbelic committed:
self-review: disregard visit notation as RootFlags is not a struct

- root flags is not a struct that we can add the dump method onto
- we also want to be able to incremently add and integrate the
serialization of root elements

this commit changes the approach to just use std::get_if so that we can
add the tests incrementally and we can revisit using visit when fully
completed

AaronBallman commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -5311,6 +5313,90 @@ void Parser::ParseMicrosoftUuidAttributeArgs(ParsedAttributes &Attrs) {
+  }
+
+  // Construct our identifier
+  StringRef Signature = StrLiteral.value()->getString();
```
Sorry for the late post-commit review, but I just noticed this patch because it's doing a lot of sema work from within the parser. Is there a reason this wasn't done using an `ActOnWhatever` function in Sema? It seems to be the only use of `PushOnScopeChains` in the parser, which is how I noticed this. CC @llvm-beanz

inbelic replied:
I have created an issue to track this here: https://github.com/llvm/llvm-project/issues/142834.

In the issue is the proposed solution and we can have the discussion there about correctness.
